### PR TITLE
HIVE-25826: Support table default types at each database level

### PR DESCRIPTION
**Pull request by saihemanth-cloudera**
### What changes were proposed in this pull request?
Introduced a new feature to support table default types at each database level

### Why are the changes needed?
Users now would be able to specify the default table type at database level. Previously setting the default table type is only possible from hive-site.xml or at table level. 

### Does this PR introduce _any_ user-facing change?
Yes. Users would need to create/alter database with the database property "defaultTableType"="EXTERNAL"

### How was this patch tested?
Local machine, remote cluster, Added unit test cases

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 1]
```diff
@@ -13600,7 +13616,12 @@ ASTNode analyzeCreateTable(
 
     // Handle different types of CREATE TABLE command
     // Note: each branch must call addDbAndTabToOutputs after finalizing table properties.
-
+    Database database  = getDatabase(qualifiedTabName.getDb());
+    boolean isDefaultTableTypeChanged = false;
+    if(database.getParameters() != null && database.getParameters().containsKey(DEFAULT_TABLE_TYPE) && database.getParameters().get(DEFAULT_TABLE_TYPE).equalsIgnoreCase("external")) {
```
Comment (nrg4878) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13621]: we should be able to compress the last 2 conditions using "getOrDefault()" call.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 2]
```diff
@@ -13600,7 +13616,12 @@ ASTNode analyzeCreateTable(
...
+      isExt = true;
+      isDefaultTableTypeChanged = true;
```
Comment (nrg4878) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13623]: appears the logic when "defaultTableType=ACID" is not implemented. I do not see where we check for this and convert to ACID automatically.

**Review by nrg4878 — CHANGES_REQUESTED** [review]

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.FileUtil;
+import org.apache.hadoop.fs.LocatedFileStatus;
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java]: Ack

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+ */
+public class TestDatabaseTableDefault {
+    static final private Logger LOG = LoggerFactory.getLogger(TestDatabaseTableDefault.class);
+
+    private HiveConf hiveConf;
+    private IDriver d;
+    private String database_with_default_table_type = "database_with_default_table_type";
+    private String default_db = "default_db";
+    private String table_type_managed = "MANAGED_TABLE";
+    private String table_type_external = "EXTERNAL_TABLE";
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java]: Ack

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        hiveConf.set(HiveConf.ConfVars.HIVEDEFAULTMANAGEDFILEFORMAT.varname, "ORC");
+        hiveConf.setVar(HiveConf.ConfVars.HIVE_TXN_MANAGER, DbTxnManager.class.getName());
+        hiveConf.set(HiveConf.ConfVars.METASTORE_CLIENT_CAPABILITIES.varname, "HIVEFULLACIDREAD,HIVEFULLACIDWRITE,HIVECACHEINVALIDATE,HIVEMANAGESTATS,HIVEMANAGEDINSERTWRITE,HIVEMANAGEDINSERTREAD");
+        TestTxnDbUtil.setConfValues(hiveConf);
+        TestTxnDbUtil.prepDb(hiveConf);
+
+        SessionState.start(new SessionState(hiveConf));
+        d = DriverFactory.newDriver(hiveConf);
+
+        conf = MetastoreConf.newMetastoreConf();
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java]: Ack

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        MetastoreConf.setBoolVar(conf, ConfVars.HIVE_IN_TEST, false);
+        MetastoreConf.setVar(conf, ConfVars.WAREHOUSE, wh.getCanonicalPath());
+        MetastoreConf.setVar(conf, ConfVars.WAREHOUSE_EXTERNAL, ext_wh.getCanonicalPath());
+        MetastoreConf.setBoolVar(conf, MetastoreConf.ConfVars.CREATE_TABLES_AS_ACID, true);
+        client = new HiveMetaStoreClient(conf);
+
+
+        dropTables();
+        runStatementOnDriver("create database " + database_with_default_table_type +" with DBPROPERTIES('defaultTableType'='EXTERNAL')");
+        runStatementOnDriver("create database " + default_db);
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java:126]: Ack

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        LOG.info("Test execution complete:testCreateManagedTables");
+    }
+
+
+    /**
+     * Tests the TestDatabaseTableDefault.testCreateExternalTables method for creating external tables in a special database.
+     * @throws Exception If there is an error creating external tables
+     */
+    @Test
+    public void testCreateExternalTables() throws Exception {
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java]: L193-215 does that tests right?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        Table external_table_3 = client.getTable(database_with_default_table_type, TableNames.EXTERNALTABLE3.toString());
+        assertEquals("Created table type is expected to be managed but found to be external", external_table_3.getTableType(), table_type_external);
+
+        runStatementOnDriver("create table " + TableNames.EXTERNALTABLE4 + " as select * from " + TableNames.EXTERNALTABLE2);
+        Table external_table_4 = client.getTable(database_with_default_table_type, TableNames.EXTERNALTABLE4.toString());
+        assertEquals("Created table type is expected to be managed but found to be external", external_table_4.getTableType(), table_type_external);
+
+        LOG.info("Test execution complete:testCreateExternalTables");
+    }
+
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java:217]: Ack

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 1 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: Ack

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 2 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: Ack. Addressed the comment in next commit

File: standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift [hunk 9]
```diff
@@ -3106,4 +3106,5 @@ const string JDBC_CONFIG_PREFIX = "hive.sql.",
 const string TABLE_IS_CTAS = "created_with_ctas",
 const string PARTITION_TRANSFORM_SPEC = "partition_transform_spec",
 const string NO_CLEANUP = "no_cleanup",
-const string CTAS_LEGACY_CONFIG = "create_table_as_external",
\ No newline at end of file
+const string CTAS_LEGACY_CONFIG = "create_table_as_external",
+const string DEFAULT_TABLE_TYPE = "defaultTableType",
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift:3139]: Used in SemanticAnalyzer#l26,L13629 e.t.c

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 10]
```diff
@@ -13615,7 +13631,22 @@ ASTNode analyzeCreateTable(
 
     // Handle different types of CREATE TABLE command
     // Note: each branch must call addDbAndTabToOutputs after finalizing table properties.
-
+    Database database  = getDatabase(qualifiedTabName.getDb());
+    boolean isDefaultTableTypeChanged = false;
+    if(database.getParameters() != null) {
+      String defaultTableType = database.getParameters().getOrDefault(DEFAULT_TABLE_TYPE, "none");
```
Comment (nrg4878) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13637]: not sure why we are using the default of "none" if not set. We are not doing anything with the value. why not just allow it to return a null when not set, and check for null?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 11]
```diff
@@ -13625,7 +13656,8 @@ ASTNode analyzeCreateTable(
       }
       tblProps = validateAndAddDefaultProperties(
           tblProps, isExt, storageFormat, dbDotTab, sortCols, isMaterialization, isTemporary,
-          isTransactional, isManaged);
+          isTransactional, isManaged, new String[] {qualifiedTabName.getDb(), qualifiedTabName.getTable()}, isDefaultTableTypeChanged);
```
Comment (nrg4878) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13659]: its not clear where the value for isDefaultTableTypeChanged is coming from in analyzeCreateTable method. The "View file" only shows about 12k lines (a truncated version of this file)

Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13659]: This flag is used as one of the determining factors in the validateAndAddDefaultProperties() to decide whether to create an acid table or managed table.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 10 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13637]: Ack

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 12]
```diff
@@ -13245,9 +13247,23 @@ private void updateDefaultTblProps(Map<String, String> source, Map<String, Strin
...
+      addDbAndTabToOutputs(qualifiedTabName,
+              TableType.MANAGED_TABLE, isTemporaryTable, retValue, storageFormat);
+    } else {
+      addDbAndTabToOutputs(qualifiedTabName,
+              TableType.EXTERNAL_TABLE, isTemporaryTable, retValue, storageFormat);
+    }
     return retValue;
   }
 
+  private boolean isExternalTableChanged (Map<String, String> tblProp, boolean isTransactional, boolean isExt, boolean isTableTypeChanged) {
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13275]: ack

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+        MetastoreConf.setVar(hiveConf, ConfVars.METASTORE_METADATA_TRANSFORMER_CLASS,
+                "org.apache.hadoop.hive.metastore.MetastoreDefaultTransformer");
+        MetastoreConf.setBoolVar(hiveConf, ConfVars.HIVE_IN_TEST, false);
+        MetastoreConf.setVar(hiveConf, ConfVars.WAREHOUSE, wh.getCanonicalPath());
+        MetastoreConf.setVar(hiveConf, ConfVars.WAREHOUSE_EXTERNAL, ext_wh.getCanonicalPath());
+        MetastoreConf.setBoolVar(hiveConf, MetastoreConf.ConfVars.CREATE_TABLES_AS_ACID, true);
+        client = new HiveMetaStoreClient(hiveConf);
+
+
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java:123]: ack

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+        dropTables();
+        runStatementOnDriver("create database " + database_with_default_table_type +" with DBPROPERTIES('defaultTableType'='EXTERNAL')");
+        runStatementOnDriver("create database " + default_db);
+    }
+
+    /**
+     * this is to test differety types of Acid tables
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java:130]: Ack

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.hadoop.hive.ql.lockmgr.DbTxnManager;
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestDatabaseTableDefault.java:44]: Ack

**Review by nrg4878 — APPROVED** [review]
Change looks good to me. +1

**Comment by nrg4878** [comment]
Fix has been merged to master. Please close the PR

**Pull request closed without merging**